src/frontend/src/handler/create_source.rs

```diff
@@ -797,42 +837,97 @@ pub(crate) fn bind_all_columns(
     sql_defined_pk_names: Vec<String>,
     with_properties: &HashMap<String, String>,
 ) -> Result<Vec<String>> {
     let sql_defined_pk = !sql_defined_pk_names.is_empty();
+    let key_column_name: Option<String> = {
+        // iter columns to check if contains additional columns from key part
+        // return the key column names if exists
+        columns.iter().find_map(|catalog| {
+            if catalog.column_desc.additional_column_type == AdditionalColumnType::Key {
+                Some(catalog.name().to_string())
+            } else {
+                None
+            }
+        })
+    };
+    let additional_column_names = columns
+        .iter()
+        .filter_map(|col| {
+            if col.column_desc.additional_column_type != AdditionalColumnType::Unspecified {
+                Some(col.name().to_string())
+            } else {
+                None
+            }
+        })
+        .collect_vec();
 
     let res = match (&source_schema.format, &source_schema.row_encode) {
         (Format::Native, Encode::Native) | (Format::Plain, _) => sql_defined_pk_names,
+
+        // For all Upsert formats, we only accept one and only key column as primary key.
+        // Additional KEY columns must be set in this case and must be primary key.
         (Format::Upsert, Encode::Json) => {
-            if sql_defined_pk {
+            if let Some(ref key_column_name) = key_column_name && sql_defined_pk {
+                if sql_defined_pk_names.len() != 1 {
+                    return Err(RwError::from(ProtocolError(
+                        format!("upsert json supports only one primary key column ({}).", key_column_name)
+                    )));
+                }
+                // the column name have been converted to real value in `handle_addition_columns`
+                // so we don't ignore ascii case here
+                if !key_column_name.eq(sql_defined_pk_names[0].as_str()) {
+                    return Err(RwError::from(ProtocolError(format!(
+                        "upsert json's key column {} not match with sql defined primary key {}",
+                        key_column_name, sql_defined_pk_names[0]
+                    ))));
+                }
                 sql_defined_pk_names
             } else {
-                add_default_key_column(columns);
-                vec![DEFAULT_KEY_COLUMN_NAME.into()]
+                return if key_column_name.is_none() {
+                    Err(
+                        RwError::from(ProtocolError("INCLUDE KEY clause must be set for FORMAT UPSERT ENCODE JSON".to_string()))
+                    )
+                } else {
+                    Err(RwError::from(ProtocolError(format!(
+                        "Primary key must be specified to {} when creating source with FORMAT UPSERT ENCODE JSON",
+                        key_column_name.unwrap()))))
+                }
             }
         }
         (Format::Upsert, Encode::Avro) => {
```

Comment:
Can we merge the `Format::Upsert, Encode::Avro` and `Format::Upsert, Encode::Json` branch because these logic are same?

Reply:
For now, yes. But we are going to support key encode, we need to derive the key structure here,whose logic is different.

Reply:
Yes, but I think it will be another match with the key encode :thinking: I prefer just match the Format in this PR

Reply:
then let's merge `format upsert encode json` and `format upsert encode avro`